qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/DateFieldsIT.java

```diff
@@ -0,0 +1,195 @@
+ */
+public class DateFieldsIT extends AbstractRollingTestCase {
+    private static final Version UPGRADE_FROM_VERSION =
+        Version.fromString(System.getProperty("tests.upgrade_from_version"));
+
+    private static final String V_6_8_1_PLUS_WARNING = "'Y' year-of-era should be replaced with 'y'. Use 'Y' for week-based-year.; " +
+        "'Z' time zone offset/id fails when parsing 'Z' for Zulu timezone. Consider using 'X'." +
+        " Prefix your date format with '8' to use the new specifier.";
+
+    private static final String V_7_0_0_PLUS_WARNING = "'Y' year-of-era should be replaced with 'y'. Use 'Y' for week-based-year.; " +
+        "'Z' time zone offset/id fails when parsing 'Z' for Zulu timezone. Consider using 'X'. " +
+        "Use new java.time date format specifiers.";
+
+    private static final String V_6_8_0_WARNING = "Use of 'Y' (year-of-era) will change to 'y' in the next major version of Elasticsearch. "
+        + "Prefix your date format with '8' to use the new specifier.";
+
+    @BeforeClass
+    public static void init(){
+        assumeTrue("upgrading from 7.0-7.6 will fail parsing joda formats",
+            UPGRADE_FROM_VERSION.before(Version.V_7_0_0));
+    }
+
+    public void testJodaBackedDocValueAndDateFields() throws Exception {
+
+        switch (CLUSTER_TYPE) {
+            case OLD:
+                Request createTestIndex = indexWithDateField("joda_it", "YYYY-MM-dd'T'HH:mm:ssZZ");
+
+                Version minVersion = getMinVersion();
+                if (minVersion.equals(Version.V_6_8_0)) {
+                    createTestIndex.setOptions(expectWarnings(V_6_8_0_WARNING));
+                } else if (minVersion.onOrAfter(Version.V_6_8_1) && minVersion.before(Version.V_7_0_0)) {
+                    createTestIndex.setOptions(expectWarnings(V_6_8_1_PLUS_WARNING));
+                } else {
+                    createTestIndex.setOptions(expectWarnings(V_7_0_0_PLUS_WARNING));
+                }
+
+                Response resp = client().performRequest(createTestIndex);
+                assertEquals(200, resp.getStatusLine().getStatusCode());
+
+                postNewDoc("joda_it/_doc");
+
+                break;
+            case MIXED:
+                postNewDoc("joda_it/_doc");
+
+                Request search = dateRangeSearch("joda_it/_search");
+
+                RequestOptions options = expectVersionSpecificWarnings(
+                    consumer -> consumer.compatible(V_6_8_0_WARNING, V_6_8_1_PLUS_WARNING, V_7_0_0_PLUS_WARNING));
+
+                search.setOptions(options);
+
+                Response searchResp = client().performRequest(search);
+                assertEquals(200, searchResp.getStatusLine().getStatusCode());
+                break;
+            case UPGRADED:
+                postNewDoc("joda_it/_doc");
+                search = dateRangeSearch("joda_it/_search");
+                //somehow  this can at times not have a warning..
```

Comment:
This sounds like a bug? It should probably warrant an issue?

Reply:
I think this was failing because I was not distinguishing between v6.8.1+ and v7  warning messages. The test was failing on warning assertion, however it is not as explicit about what was wrong as the yml assertion (which tells what was actual and what expected)
I can't reproduce at the moment, but will keep an eye on it and create an issue/fix when it occurs

Reply:
also I was often unaware that the test was not running.. https://github.com/elastic/elasticsearch/pull/52555#discussion_r387090151

Reply:
ok this is odd. it is failing on https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request-bwc/17251/
but it is passing for me now locally (failing as expected too). 
@polyfractal Could I somehow not be hitting the DocValue date logic with just sorting?

Reply:
To be honest I'm not sure, sorting has a slightly different mechanism that goes through the search infrastructure.  If you did an aggregation with a `date_histogram` and it returns at least one bucket, you'd be guaranteed to serialize a bucket with the date formatter.

Reply:
I also discussed this with @jpountz and we confirmed that there is a chance when this logic won't be used. 
In rolling upgrade tests ES might decide to allocate all shards to older nodes in mixed cluster and in upgraded cluster it might still remain on single node. Client might then connect to this node with all shards, therefore the serialisation logic won't be used.

I iterate 3 times to make sure client at some point connects to a node that do not contain shards for these indices and triggers serialisation logic.